scapy/fields.py

```diff
@@ -318,7 +318,17 @@ def template(self, pkt, val):
             return getattr(self._find_fld_pkt_val(pkt, val), func)(pkt, val)
         return template
 
-    any2i = _wrapper_fields_func("any2i")
+    # any2i is called when building a packet with Packet(stuff=value)
+    # In that case, the other fields are not yet builded when it gets called,
+    # meaning that the conditions cannot work yet. So the first inital value
+    # must not be field-specific.
+    #  e.g.: in SOCKS:
+    #        SOCKS5Request(atyp=0x3, addr="scapy.net")
+    #  we want addr to be handled by the DNS module, but started like so, if
+    #  any2i was called in the first place, it would return Net("scapy.net")
+    #  as the default field is an IP, which would have led to a crash later
+    # SOLUTION: disable it by using a dummy lambda
+    any2i = lambda self, pkt, x: x
```

Comment:
@p-l- I have an issue with MultipleTypeField,
this patch does not work, but it explains the issue.

Basically, in this PR i use this field:

```
    MultipleTypeField(
        [
            # IPv4
            (IPField("addr", "0.0.0.0"), lambda pkt: pkt.atyp == 0x1),
            # DNS
            (DNSStrField("addr", ""), lambda pkt: pkt.atyp == 0x3),
            # IPv6
            (IP6Field("addr", "::"), lambda pkt: pkt.atyp == 0x4),
        ],
        StrField("addr", "")
    ),
```

Which is instanciated with, for instance

`SOCKS5Request(atyp=0x3, addr="scapy.net")`

If I get the fields's value, I will get `Net("scapy.net")`

Reply:
I was not able to fix it, after several trials :/ Maybe you have an idea

Reply:
Found a way